### ISSUE TITLE
Add Flux CRDs in OLM ClusterServiceVersion

### ISSUE THREAD

**Issue by blezoray** [controlplaneio-fluxcd/flux-operator#61]
Hello,

With Openshift (4.14), in release 0.6.0, flux-operator ClusterServiceVersion contains only `FluxInstance`and `FluxReport` CRDs that it manages.

But, for endusers, the console shows only these 2 kind of resources:
![image](https://github.com/controlplaneio-fluxcd/flux-operator/assets/17479277/d167c9de-98e6-4f84-b3a2-29275367aa08)

In order to let the console showing endusers flux resources, like GitRepository ..., all the flux CRDs must be added to the ClusterServiceVersion even if the `flux-operator` controller doesn't manage them.

Could you add them to the CSV ?

BR.

**Comment by stefanprodan** [comment]
Why would anyone create Flux resources from a UI? All of these should be in Git

**Comment by stefanprodan** [comment]
Ok so this is about a readonly view, I get it now. 

Does OLM allows adding CRDs to the owned list even if the operator does not "owns" them?

**Comment by blezoray** [comment]
Yes, it is the case for External Secrets Operator, based on helm operator SDK. There is the `OperatorConfig` equivalent to `FluxInstance`

The consequence is that flux-operator must not install CRDs

**Comment by stefanprodan** [comment]
Wouldn't the UI crash if some CRD let's say `HelmRelease` is listed in the CVS but is not installed in the cluster?

**Comment by stefanprodan** [comment]
After adding the CRDs the bundle no longet builds as see in #62

```
time="2024-07-03T05:51:49Z" level=error msg="permissive mode disabled" bundles="[ghcr.io/controlplaneio-fluxcd/openshift-flux-operator-catalog:bundle-0.6.0]" error="error checking provided apis in bundle flux-operator.v0.6.0: couldn't find toolkit.fluxcd.io/v1beta2/Bucket (source) in bundle. found: map[fluxcd.controlplane.io/v1/FluxInstance (fluxinstances):{} fluxcd.controlplane.io/v1/FluxReport (fluxreports):{}]"
```

**Comment by stefanprodan** [comment]
Yeah I don't think this can work, Operator SDK wants all CRDs inside the bundle:

```console
$ operator-sdk bundle validate .
ERRO[0000] Error: Value toolkit.fluxcd.io/v1, Kind=GitRepository: owned CRD "toolkit.fluxcd.io/v1, Kind=GitRepository" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1, Kind=HelmRepository: owned CRD "toolkit.fluxcd.io/v1, Kind=HelmRepository" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1, Kind=HelmChart: owned CRD "toolkit.fluxcd.io/v1, Kind=HelmChart" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta2, Kind=Bucket: owned CRD "toolkit.fluxcd.io/v1beta2, Kind=Bucket" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta2, Kind=OCIRepository: owned CRD "toolkit.fluxcd.io/v1beta2, Kind=OCIRepository" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1, Kind=Kustomization: owned CRD "toolkit.fluxcd.io/v1, Kind=Kustomization" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v2, Kind=HelmRelease: owned CRD "toolkit.fluxcd.io/v2, Kind=HelmRelease" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta3, Kind=Provider: owned CRD "toolkit.fluxcd.io/v1beta3, Kind=Provider" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta3, Kind=Alert: owned CRD "toolkit.fluxcd.io/v1beta3, Kind=Alert" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1, Kind=Receiver: owned CRD "toolkit.fluxcd.io/v1, Kind=Receiver" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta2, Kind=ImageRepository: owned CRD "toolkit.fluxcd.io/v1beta2, Kind=ImageRepository" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta2, Kind=ImagePolicy: owned CRD "toolkit.fluxcd.io/v1beta2, Kind=ImagePolicy" not found in bundle "flux-operator.v0.6.0" 
ERRO[0000] Error: Value toolkit.fluxcd.io/v1beta2, Kind=ImageUpdateAutomation: owned CRD "toolkit.fluxcd.io/v1beta2, Kind=ImageUpdateAutomation" not found in bundle "flux-operator.v0.6.0" 

```

**Comment by blezoray** [comment]
I think you need to add also the CRD in the manifest directory, like for FluxInstance and FluxReport.

**Comment by stefanprodan** [comment]
That's not an option, it would break the whole functionality of Flux Operator.

**Comment by blezoray** [comment]
Why ?
This is the case for External Secrets Operator:
https://github.com/redhat-openshift-ecosystem/community-operators-prod/tree/main/operators/external-secrets-operator/0.9.19/manifests

**Comment by stefanprodan** [comment]
Flux is not External Secrets Operator, Flux is composable, for example on a production cluster you would not want the image automation CRDs nor controllers. Also the Flux Operator is meat to install different version of Flux, which come with different versions of CRDs, if the CRDs are in bundle, the Flux Operator will fight with OLM over those resources.

**Comment by blezoray** [comment]
Except if you have admission or mutating webhook, there is no impact on a CustomResource if there is no controller that manages it. The resource will be created in the ETCD and that's all

**Comment by stefanprodan** [comment]
The impact is to Flux, if you set `version: v2.2.x` and the CRDs are from v2.3 all controllers will crash loop, also the automated upgrade will fail, as OLM will undo the CRDs if Flux Operator upgrades them to let's say a future v2.4

**Comment by blezoray** [comment]
in this case, you should manage different versions of CRDs, v1beta1, v1beta2, ...

**Comment by stefanprodan** [comment]
We do, but that's not as easy it sounds, Kubernetes doesn't allow CRDs downgrade and the default conversion webhook for CRs will bump the version in etcd only at the first mutation. The whole goal of Flux Operator is to manage the CRDs/CRs upgrade for the user, if the CRDs are applied at some version set in OLM we are back where the Weaveworks operator was and I have no intention of doing that.

**Comment by stefanprodan** [comment]
OLM can't remove a Flux version at all, because finalizers will block it, some managed workloads will be deleted and the whole cluster will be stuck. The Flux Operator knows how to uninstall Flux properly without affecting any workloads deployed on the cluster. So with the operator, if you want to downgrade from a minor to another, you can by simply deleting the FluxInstance and create a new one, while no workloads will be touched.

**Comment by stefanprodan** [comment]
Maybe there is some programmatic way of telling OpenShift which CRDs are available, if there is such an API then the operator would call it during the report generation.

**Comment by blezoray** [comment]
> Maybe there is some programmatic way of telling OpenShift which CRDs are available, if there is such an API then the operator would call it during the report generation.

I don't think so. the CRDs are statically described and if you change them through the API, OLM would overwrite them.

Is there many changes in the flux CRDs during the last 2 or 3 years that justify you need to manager CRDs at controller level ?

**Comment by stefanprodan** [comment]
> Is there many changes in the flux CRDs during the last 2 or 3 years that justify you need to manager CRDs at controller level ?

Every single Flux minor comes with changes to CRDs, Flux evolves along with its APIs

**Comment by blezoray** [comment]
Therefore, the operator should have a new version. I don't see another alternative
